fuzzing/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+//! This is the 'timelock' example modified slightly to demonstrate
+//! Soroban's fuzzing capabilities.
+//!
+//! This contract demonstrates 'timelock' concept and implements a
+//! greatly simplified Claimable Balance (similar to
+//! https://developers.stellar.org/docs/glossary/claimable-balance).
+//! The contract allows to deposit some amount of token and allow another
+//! account(s) claim it before or after provided time point.
+//! For simplicity, the contract only supports invoker-based auth.
+
+#![no_std]
+
+use soroban_sdk::{contract, contractimpl, contracttype, token, Address, Env, Vec};
+
+#[derive(Clone)]
+#[contracttype]
+pub enum DataKey {
+    Init,
+    Balance,
+}
+
+#[derive(Clone, Debug)]
+#[contracttype]
+pub enum TimeBoundKind {
+    Before,
+    After,
+}
+
+#[derive(Clone, Debug)]
+#[contracttype]
+pub struct TimeBound {
+    pub kind: TimeBoundKind,
+    pub timestamp: u64,
+}
+
+#[derive(Clone)]
+#[contracttype]
+pub struct ClaimableBalance {
+    pub token: Address,
+    pub amount: i128,
+    pub claimants: Vec<Address>,
+    pub time_bound: TimeBound,
+}
+
+#[contract]
+pub struct ClaimableBalanceContract;
+
+// The 'timelock' part: check that provided timestamp is before/after
+// the current ledger timestamp.
+fn check_time_bound(env: &Env, time_bound: &TimeBound) -> bool {
+    let ledger_timestamp = env.ledger().timestamp();
+
+    match time_bound.kind {
+        TimeBoundKind::Before => ledger_timestamp <= time_bound.timestamp,
+        TimeBoundKind::After => ledger_timestamp >= time_bound.timestamp,
+    }
+}
+
+#[contractimpl]
+impl ClaimableBalanceContract {
+    pub fn deposit(
+        env: Env,
+        from: Address,
+        token: Address,
+        amount: i128,
+        claimants: Vec<Address>,
+        time_bound: TimeBound,
+    ) {
+        // Perhaps this check should be enabled...
+        /*if amount == 0 {
```

Comment:
We should check that `amount` is negative at a minimum since it's a `i128`.

Reply:
I don't believe it is necessary to check for a negative amount. The claimable balance in this contract never goes below zero, and I think the reason is that the token contract disallows negative balance transfers: https://github.com/stellar/rs-soroban-env/blob/main/soroban-env-host/src/native_contract/token/contract.rs#L214

This contract is copied directly from the pre-existing timelock example, so any changes here are probably needed there as well.

This commented out block was added by me as extra logic that I think is reasonable (claiming a zero balance is arguably not useful), and is checked by the accompanying fuzz tests so that the fuzz tests will fail. The user following the accompanying tutorial is expected to uncomment this line to make the fuzzer succeed.

Reply:
While the SAC and token example check for negative amounts, this technically doesn't have to be true. Someone can implement a token that accepts negative values. Good point about the timelock contract though. I'll merge this for now and open an issue about revisiting instances like this.